src/Education/XamlGenerator/RawTextProcessor.cs

```diff
@@ -0,0 +1,92 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text.RegularExpressions;
+
+namespace SonarLint.VisualStudio.Education.XamlGenerator
+{
+    /// <summary>
+    /// Processes special internal markup embedded in normal text
+    /// </summary>
+    internal class RawTextProcessor
+    {
+        /// <summary>
+        /// Processes a text string that might/might not contain rule cross-references
+        /// </summary>
+        public static List<IRawTextItem> ProcessRawText(string rawText)
+        {
+            var rawTextItems = new List<IRawTextItem>();
+
+            // Matches -   ... text ... {rule:cpp:S123} ... text ... {rule:cpp:S999} ... text
+            var matches = Regex.Matches(rawText,
+                "{rule:([A-Za-z]+):([\\w]+)}");
+
+            int endOfLastMatch = 0;
+
+            foreach (Match match in matches)
+            {
+                Debug.Assert(match.Success, "Expecting match to have succeeded");
+                var matchGroup = match.Groups[0];
+
+                var simpleText = rawText.Substring(endOfLastMatch, matchGroup.Index - endOfLastMatch);
+                AddSimpleText(rawTextItems, simpleText);
+
+                rawTextItems.Add(new RawTextItem(matchGroup.Value, true));
```

Comment:
The matchGroup.Value will simply be -> "{rule:cpp:S123}"

Reply:
Is the caller going to be able to use this value directly, or are they going to have to do some more parsing to turn it into a usable form? e.g. stripping off the curly brackets, and perhaps extracting the composite rule id?

Ideally, the caller just wants to know that this is a cross reference and what the referenced composite rule id is without having to do any extra work.

Looking at it another way, the parser is the only thing that should need to know exactly how the cross-references are encoded in the raw text. If the format of that encoding changed (to e.g. `[CROSSREF repo=cs;rule=S123]`), only the parser should need to change.

Reply:
Hm those are good points, I was thinking the parse doesn't know what to actually do with the information so it should leave it whole but the way you explain it makes more sense